sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -1447,3 +1450,133 @@ object EmbedSerializerInFilter extends Rule[LogicalPlan] {
+
+  /**
+   * Prepare an [[InSubQuery]] by rewriting it (in case of correlated predicates) and by
+   * constructing the required join condition. Both the rewritten subquery and the constructed
+   * join condition are returned.
+   */
+  private def rewriteInSubquery(
+      subquery: InSubQuery,
+      query: LogicalPlan): (LogicalPlan, Expression) = {
+    val expressions = subquery.expressions
+    val (resolved, joinCondition) = pullOutCorrelatedPredicates(subquery.query, query)
+    val conditions = joinCondition.toSeq ++ expressions.zip(resolved.output).map(EqualTo.tupled)
+    (resolved, conditions.reduceLeft(And))
+  }
+
+  def apply(plan: LogicalPlan): LogicalPlan = plan transform {
+    case f @ Filter(condition, child) =>
+      val (withSubquery, withoutSubquery) =
+        splitConjunctivePredicates(condition).partition(PredicateSubquery.hasPredicateSubquery)
+
+      // Construct the pruned filter condition.
+      val newFilter: LogicalPlan = withoutSubquery match {
+        case Nil => child
+        case conditions => Filter(conditions.reduce(And), child)
+      }
+
+      // Filter the plan by applying left semi and left anti joins.
+      withSubquery.foldLeft(newFilter) {
+        case (p, Exists(sub)) =>
+          val (resolved, joinCondition) = pullOutCorrelatedPredicates(sub, p)
+          Join(p, resolved, LeftSemi, joinCondition)
+        case (p, Not(Exists(sub))) =>
+          val (resolved, joinCondition) = pullOutCorrelatedPredicates(sub, p)
+          Join(p, resolved, LeftAnti, joinCondition)
+        case (p, in: InSubQuery) =>
+          val (resolved, cond) = rewriteInSubquery(in, p)
+          Join(p, resolved, LeftSemi, Option(cond))
+        case (p, Not(in: InSubQuery)) =>
+          // This is a NULL-aware (left) anti join (NAAJ).
+          // We currently only allow subqueries with non-nullable fields. In this way we can plan a
+          // regular ANTI join, instead of a much more complex NAAJ (which is not yet available in
+          // Spark SQL). In order to get the NAAJ semantically right, we need to add a filter to
+          // left hand side of the query that checks that either all columns are non-null or that
+          // the right hand side is empty.
+          val (resolved, cond) = rewriteInSubquery(in, p)
+
+          // Make absolutely sure that the rewritten query contains no nullable fields. We re-check
+          // this here because the rewritten query can contain pulled-up nullable columns.
+          if (resolved.output.exists(_.nullable)) {
+            throw new AnalysisException("NOT IN with nullable subquery is not supported. " +
+              "Please use a non-nullable sub-query or rewrite this using NOT EXISTS.")
+          }
+
+          // Construct filter for the left hand side
+          val count = Alias(AggregateExpression(Count(Literal(1)), Complete, false), "cnt")()
+          val isEmpty = EqualTo(ScalarSubquery(Aggregate(Nil, Seq(count), resolved)), Literal(0L))
```

Comment:
I do like the idea of adding a `Semi Join Plus`, it would solve a lot and is a logical next step.

The stopping criteria for a SEMI JOIN and a NAAJ would have to be slightly different though:

- As soon as we find a match in a SEMI JOIN we are done.
- In an NAAJ we have to check for both a match and a partial match (one or more NULLs).